### ISSUE TITLE
default is not permitted as constant pattern with nullable types

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#20015]
**Version Used**: master

**Steps to Reproduce**:

```cs
bool a = default(int?) is default(int?);  // error CS0150: A constant value is expected
bool b = default(int?) is default;        // error CS0150: A constant value is expected
```

**Expected Behavior**: No error

**Actual Behavior**: A constant value is expected

**Comment by jcouv** [comment]
FYI @gafter for triage

**Comment by gafter** [comment]
The expression on the right-hand-side of the `is` operator is not "target typed", and therefore must be an expression that has a type. This behavior is by design.

**Comment by alrz** [comment]
two things: 

This is an error even with a typed `default(int?)`
`default` works for an `object`: `obj is default`

**Comment by gafter** [comment]
Hmmm. Will investigate more.

**Comment by gafter** [comment]
According to the language specification, a value of type `int?` is not a constant.

**Comment by gafter** [comment]
I expect it is likely we will want to forbid patterns of the form `default`, to reduce confusion when we introduce the switch expression in the future. As such, I think it would be counterproductive to fix this.

**Comment by alrz** [comment]
Nothing about `is default` is confusing, this is just to make it consistent with itself (this is the only case that is not permitted). switch statement already produce a warning for the direct use (but not `(default)`, for instance). I think switch expression could follow the same rule, to be not confused with the "default" case.

**Comment by sharwell** [comment]
💭 I would expect both of the cases in the original example to behave like this:

```csharp
bool c = default(int?) is null;
```

**Comment by jcouv** [comment]
I'll move to csharplang, since currently by design.

**Comment by jcouv** [comment]
Issue moved to [dotnet/csharplang #1135](https://github.com/dotnet/csharplang/issues/1135) via [**ZenHub**](https://www.zenhub.com/)
